clients/drcachesim/drptpostprocessor/drptpostprocessor.c

```diff
@@ -0,0 +1,649 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* DrPTPostprocessor.c
+ * command-line tool for decoding a PT trace, and converting it into an instruction-only
+ * memtrace composed of 'memref_t's
+ * XXX: Currently, it only counts and prints the instruction count in the trace data.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <inttypes.h>
+#include <errno.h>
+#include <linux/limits.h>
+
+#include "load_elf.h"
+#include "pt_cpu.h"
+#include "intel-pt.h"
+#include "libipt-sb.h"
+
+#define IN
+#define OUT
+#define INOUT
+
+/* DR IntelPT decoder*/
+typedef struct _dript_decoder_t {
+    /* libipt configuration */
+    struct pt_config config;
+
+    /* libipt instruction decoder*/
+    struct pt_insn_decoder *instr_decoder;
+
+    /* The image section cache. */
+    struct pt_image_section_cache *iscache;
+
+    /* The preload image. */
+    struct pt_image *preload_image;
+
+    /* The perf event sideband decoder configuration. */
+    struct pt_sb_pevent_config sb_pevent_conf;
+
+    /* The sideband session. */
+    struct pt_sb_session *sb_session;
+} dript_decoder_t;
+
+/* A collection of options. */
+typedef struct _dript_options_t {
+    /* Print statistics. */
+    uint32_t print_stats : 1;
+} dript_options_t;
+
+/* A collection of output. */
+typedef struct _dript_output_t {
+    /* The number of instructions. */
+    uint64_t instr_count;
+} dript_output_t;
+
+/* Diagnostic output for error event.
+ * It will be called when 'pt_insn_sync_forward' return error status or decoder end before
+ * reach pte_eos.
+ */
+static void
+diagnose_error(struct pt_insn_decoder *decoder IN, int errcode IN, const char *errtype IN,
+               uint64_t ip IN)
+{
+    int err;
+    uint64_t pos;
+
+    err = -pte_internal;
+    pos = 0ull;
```

Comment:
Do we need to init this to 0? Given that we don't use it if `pt_insn_get_offset` returns an error.